1.20 - 1.21.1 with even more Glue

Fourmisain opened the pull request:
**Edit**: Removed the old description, since it's mostly out of date.

See the last commit message and of course everything below here for details.

Fourmisain committed:
update to loom 1.10, gradle 8.13

Fourmisain committed:
enable all build warnings

Fourmisain committed:
- update to Glue 0.0.6
- restructure the project so every file path is unique per subproject
- handle per-project mixin names (so they don't have to be globally unique)
- add asserts to ensure we don't miss a version range
- add 1.20, 1.20.3 and 1.20.5 MixinClientTitleScreen similar to #21
- update mixin plugin with values from #21
- use client/server mod initializers to guarantee MinepkgCompanion run order
- squash 1_19_3 into 1_17
- bump to 0.8.0

Fourmisain commented:
Hm... looks like the mixins all need to same `compatibilityLevel` 16.

This might be a problem when 1.16 and below should be supported as then it would have to be Java 8 - but Java 8 as a release target is already deprecated for removal since Java 21, so having one build for all versions might in the future not be possible in this way anymore...

Fourmisain commented:
I thought this worked - and it mostly does, except the `CODEC` field is not remapped to `field_42535` for some reason.

Maybe Glue isn't that bad afterall...

oliwoli commented:
> I thought this worked - and it mostly does, except the `CODEC` field is not remapped to `field_42535` for some reason.
> 
> Maybe Glue isn't that bad afterall...

I think it might work when you use `import net.minecraft.text.TextCodecs` instead (starting at 1.20.3)

And in the TitleScreen these are my imports:
```
import io.minepkg.companion.MinepkgCompanion;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.screen.multiplayer.ConnectScreen;
import net.minecraft.client.gui.screen.TitleScreen;
import net.minecraft.client.network.ServerAddress;
import net.minecraft.client.network.ServerInfo;
import net.minecraft.client.network.ServerInfo.ServerType;
import net.minecraft.world.level.storage.LevelStorage;
import net.minecraft.world.level.storage.LevelStorageException;
import net.minecraft.world.level.storage.LevelSummary;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
```
And then I changed the joinServer function to this:
```
	private void joinServer (String hostname) {
    MinecraftClient client = MinecraftClient.getInstance();
    // Create a server entry
    ServerInfo entry = new ServerInfo(hostname, hostname, ServerType.OTHER); // Added ServerType.OTHER
    // Join the server
    ConnectScreen.connect(client.currentScreen, client, ServerAddress.parse(entry.address), entry, false); // Added ", false" at the end
	}
```
Also, createIntegratedServerLoader().start expects a runnable:  `client.createIntegratedServerLoader().start(level.getName(), () -> {}); // Passing empty Runnable`

Minecraft really do be trolling.

But I still couldn't get it to actually load a world in 1.20.3. Gets to 99% and then:
```
Description: Exception in server tick loop

java.lang.NoSuchFieldError: field_40722
        at knot//net.minecraft.class_2926.md44b9fd$minepkg-companion$lambda$static$2$0(class_2926.java:1020)
        at knot//com.mojang.serialization.codecs.RecordCodecBuilder.create(RecordCodecBuilder.java:72)
        at knot//net.minecraft.class_2926.<clinit>(class_2926.java:20)
        at knot//net.minecraft.server.MinecraftServer.method_49385(MinecraftServer.java:901)
        at knot//net.minecraft.server.MinecraftServer.method_29741(MinecraftServer.java:654)
        at knot//net.minecraft.server.MinecraftServer.method_29739(MinecraftServer.java:270)
        at java.base/java.lang.Thread.run(Thread.java:840)
```

Fourmisain commented:
> I think it might work when you use import net.minecraft.text.TextCodecs instead (starting at 1.20.3)

I do do that, the issue I'm having is purely a build issue, the main project doesn't actually get the finished, remapped actual final real final output (v2), so I tried remapping it in the main project, which works to 99%, but not for that one particular field apparently.
I'll see if there's a way to make this `jar` build hack work, else it's back to Glue.

> But I still couldn't get it to actually load a world in 1.20.3. Gets to 99% and then:
> NoSuchFieldError: field_40722

That is `Codecs.TEXT`, which is from `mc1_19_4`'s `ServerMetadataMixin`, you need to make sure it's not loaded in 1.20.3+ by adjusting the version range in `GlueMixinPlugin`.

Which is actually a problem in this version as I have 2 `ServerMetadataMixin`, one in `mc1_19_4` and one in `1_20_3`.
The mixin plugin only distinguishes by filename and thus uses the wrong version range.

Apparently I also didn't finish the version range thing as I have a `"ServerMetadataMixin2"` under 1.19.4, which I don't think should be there. That code was commented out when I came back to the project, so that was apparently another TODO.

oliwoli commented:
> That is `Codecs.TEXT`, which is from `mc1_19_4`'s `ServerMetadataMixin`, you need to make sure it's not loaded in 1.20.3+ by adjusting the version range in `GlueMixinPlugin`.
> 
> Which is actually a problem in this version as I have 2 `ServerMetadataMixin`, one in `mc1_19_4` and one in `1_20_3`. The mixin plugin only distinguishes by filename and thus uses the wrong version range.
> 
> Apparently I also didn't finish the version range thing as I have a `"ServerMetadataMixin2"` under 1.19.4, which I don't think should be there. That code was commented out when I came back to the project, so that was apparently another TODO.

Okay... I will investigate... tomorrow... maybe. In the meantime I pushed my wip mess [here](https://github.com/minepkg/companion-fabric/tree/1.20-wip-with-glue)

Thanks a bunch for your help and advice! :)

Fourmisain commented:
I updated Glue, getting rid of most the jank and deprecated code.
The project should now build correctly, though it still has the same issues as before, TBD another day.

oliwoli commented on .github/workflows/build-and-test.yml:
```diff
@@ -61,6 +61,18 @@ jobs:
+            1.21.1,
+            1.21.2,
+            1.21.3,
+            1.21.4
```
Don't bother supporting >=1.21.2 for now. The duplicate ASM classes crash seems to be caused by the minepkg cli. 
`minepkg launch vanilla -m 1.21.2` only includes asm v9.3 while the default launch command (even without any mods) includes asm v9.3 as well as v9.7.1.

edit: I mean feel free to make the mod compatible but we should remove the tests in the crashtest that definitely won't work.

Fourmisain commented:
I _think_ this should be it, now everything just needs to be tested.

I added the TitleScreenScreen mixins like #21 (with slight fixes), used the same versions ranges that Oli tested (at least I hope I did).
Reworked the GlueMixinPlugin to be way more readable and throw assertion errors when you forget a version range.
Also, mixin names don't need to be globally unique anymore, no more `MixinTitleScreenScreen3264`.

It's still annoying that the dev env can't be used. I could slightly tweak the project so you could at least run the subprojects, but that's not ideal either.

Right now, every file is completely unique, so Glue doesn't actually have to merge any files at all.

...

Uh oh.
What is this
![uhm](https://github.com/user-attachments/assets/ff9d1e09-bf99-4a9a-a232-bc1b4abfc60b)

The mixin config files are all empty? All except for `common`, the last project to be glued.

Glue is broken, apparently, even though it worked just a second ago...

Fourmisain commented:
Ohhh... because I removed the fabric.mod.json from the subprojects, it doesn't think there's any mixins - which is technically correct...

oliwoli commented:
Okay great! I'll test it when I get the chance and then let's get this thing merged!

Fourmisain commented:
So... good news and bad news.

Good news:
The build now works.
I re-added the fabric.mod.jsons to each subproject 73cf4504331ad0a43035bc0f43f58f144821e3dc (not sure why the timeline doesn't show that), each project now simply defines it's own mixins and nothing else and Glue combines everything like it was always supposed to.

Bad news:
I think there's a potentially unsolvable, fundamental problem with Mixin we might run into at some point.

We have a bunch of mixin configs, all using the same mixin plugin that will only apply each mixin for the appropriate Minecraft version.
Each mixin config, however, also defines a [compatibilityLevel](https://github.com/SpongePowered/Mixin/wiki/PSA-Forward-Compatibility-Features-in-Mixin#2-mixin-config-property-compatibilitylevel) which "can be used to elevate Mixin's operating level to the required level for the feature set you wish to use".

So generally, you always want to set it to the Java version of the Minecraft version the mixin targets.
But,

> if the current runtime does not support the requested feature set (eg. attempting to elevate to JAVA_9 whilst running Java 8) then the operation can fail fast

Meaning we are forced to set it to a lower level, which is currently `JAVA_16`.

I'm still not entirely sure if that means that mixin will just refuse to use certain features, even though the runtime supports them - it sure sounds like it.

E.g. the language feature "records" [was defined for `JAVA_14`](https://github.com/SpongePowered/Mixin/blob/4053421aa10aaac6127d969028a29c94fe3054f6/src/main/java/org/spongepowered/asm/mixin/MixinEnvironment.java#L742-L744) - so would a mixin targeting a record fail when the `compatibilityLevel` is set to `JAVA_8`?
I don't know.

oliwoli commented:
Test launching 1.17 (not with the try feature, just launching the game with a world):
```
[13:59:09] [main/INFO]: Loading Minecraft 1.17.1 with Fabric Loader 0.16.10
[13:59:10] [main/INFO]: Loading 6 mods:
        - fabricloader 0.16.10
           \-- mixinextras 0.4.1
        - java 16
        - minecraft 1.17.1
        - minepkg-companion 0.8.0
           \-- com_moandjiezana_toml_toml4j 0.7.2
[13:59:10] [main/INFO]: SpongePowered MIXIN Subsystem Version=0.8.7 Source=file:/home/oliwoli/.cache/minepkg/libraries/net/fabricmc/sponge-mixin/0.15.4+mixin.0.8.7/sponge-mixin-0.15.4+mixin.0.8.7.jar Service=Knot/Fabric Env=CLIENT
[13:59:10] [main/INFO]: Compatibility level set to JAVA_16
[13:59:13] [Render thread/INFO]: Initializing MixinExtras via com.llamalad7.mixinextras.service.MixinExtrasServiceImpl(version=0.4.1).
[13:59:14] [Render thread/INFO]: Environment: authHost='https://authserver.mojang.com', accountsHost='https://api.mojang.com', sessionHost='https://sessionserver.mojang.com', servicesHost='https://api.minecraftservices.com', name='PROD'
[13:59:14] [Render thread/INFO]: Setting user: ciuryburie
[13:59:14] [Render thread/INFO]: minepkg companion for Minecraft < 1.19.4 ready
[13:59:15] [Render thread/INFO]: Backend library: LWJGL version 3.2.2 build 10
[13:59:16] [Render thread/WARN]: ERROR : Couldn't load Narrator library : Unable to load library 'fliteWrapper':
libflite.so.1: cannot open shared object file: No such file or directory
libflite.so.1: cannot open shared object file: No such file or directory
...
blablabla
....
[13:59:43] [Render thread/INFO]: Preparing spawn area: 97%
[13:59:43] [Render thread/INFO]: Time elapsed: 9158 ms
[13:59:43] [Server thread/INFO]: Changing view distance to 11, from 10
[13:59:44] [Netty Local Client IO #0/FATAL]: java.lang.UnsupportedClassVersionError: net/minecraft/class_2924 has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 60.0
[13:59:44] [Netty Server IO #1/FATAL]: java.lang.NoClassDefFoundError: Could not initialize class net.minecraft.class_2539
[13:59:44] [Server thread/INFO]: Stopping server
[13:59:44] [Server thread/INFO]: Saving players
[13:59:44] [Server thread/INFO]: Saving worlds
[13:59:44] [Server thread/INFO]: Saving chunks for level 'ServerLevel[test]'/minecraft:overworld
[13:59:44] [Server thread/INFO]: Saving chunks for level 'ServerLevel[test]'/minecraft:the_nether
[13:59:44] [Server thread/INFO]: Saving chunks for level 'ServerLevel[test]'/minecraft:the_end
[13:59:44] [Server thread/INFO]: ThreadedAnvilChunkStorage (test): All chunks are saved
[13:59:44] [Server thread/INFO]: ThreadedAnvilChunkStorage (DIM-1): All chunks are saved
[13:59:44] [Server thread/INFO]: ThreadedAnvilChunkStorage (DIM1): All chunks are saved
[13:59:44] [Server thread/INFO]: ThreadedAnvilChunkStorage: All dimensions are saved
[13:59:45] [Render thread/FATAL]: Reported exception thrown!
net.minecraft.class_148: mouseClicked event handler
        at net.minecraft.class_437.method_25412(class_437.java:495) ~[client-intermediary.jar:?]
        at net.minecraft.class_312.method_1601(class_312.java:94) ~[client-intermediary.jar:?]
        at net.minecraft.class_312.method_22686(class_312.java:165) ~[client-intermediary.jar:?]
        at net.minecraft.class_1255.execute(class_1255.java:101) ~[client-intermediary.jar:?]
        at net.minecraft.class_312.method_22684(class_312.java:165) ~[client-intermediary.jar:?]
        at org.lwjgl.glfw.GLFWMouseButtonCallbackI.callback(GLFWMouseButtonCallbackI.java:36) ~[lwjgl-glfw-3.2.2.jar:?]
        at org.lwjgl.system.JNI.invokeV(Native Method) ~[lwjgl-3.2.2.jar:?]
        at org.lwjgl.glfw.GLFW.glfwWaitEventsTimeout(GLFW.java:3174) ~[lwjgl-glfw-3.2.2.jar:?]
        at com.mojang.blaze3d.systems.RenderSystem.limitDisplayFPS(RenderSystem.java:190) ~[client-intermediary.jar:?]
        at net.minecraft.class_310.method_1523(class_310.java:1140) ~[client-intermediary.jar:?]
        at net.minecraft.class_310.method_1514(class_310.java:728) [client-intermediary.jar:?]
        at net.minecraft.client.main.Main.main(Main.java:217) [client-intermediary.jar:?]
        at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:480) [fabric-loader-0.16.10.jar:?]
        at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74) [fabric-loader-0.16.10.jar:?]
        at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23) [fabric-loader-0.16.10.jar:?]
Caused by: java.lang.NoClassDefFoundError: Could not initialize class net.minecraft.class_2539
        at net.minecraft.class_310.method_29610(class_310.java:2087) ~[client-intermediary.jar:?]
        at net.minecraft.class_310.method_29606(class_310.java:1939) ~[client-intermediary.jar:?]
        at net.minecraft.class_528$class_4272.method_20174(class_528.java:437) ~[client-intermediary.jar:?]
        at net.minecraft.class_528$class_4272.method_20164(class_528.java:343) ~[client-intermediary.jar:?]
        at net.minecraft.class_528$class_4272.method_25402(class_528.java:279) ~[client-intermediary.jar:?]
        at net.minecraft.class_350.method_25402(class_350.java:359) ~[client-intermediary.jar:?]
        at net.minecraft.class_4069.method_25402(class_4069.java:27) ~[client-intermediary.jar:?]
        at net.minecraft.class_312.method_1611(class_312.java:94) ~[client-intermediary.jar:?]
        at net.minecraft.class_437.method_25412(class_437.java:490) ~[client-intermediary.jar:?]
        ... 14 more
---- Minecraft Crash Report ----
// Don't be sad, have a hug! <3

Time: 3/13/25, 1:59 PM
Description: mouseClicked event handler

java.lang.NoClassDefFoundError: Could not initialize class net.minecraft.class_2539
        at net.minecraft.class_310.method_29610(class_310.java:2087)
        at net.minecraft.class_310.method_29606(class_310.java:1939)
        at net.minecraft.class_528$class_4272.method_20174(class_528.java:437)
        at net.minecraft.class_528$class_4272.method_20164(class_528.java:343)
        at net.minecraft.class_528$class_4272.method_25402(class_528.java:279)
        at net.minecraft.class_350.method_25402(class_350.java:359)
        at net.minecraft.class_4069.method_25402(class_4069.java:27)
        at net.minecraft.class_312.method_1611(class_312.java:94)
        at net.minecraft.class_437.method_25412(class_437.java:490)
        at net.minecraft.class_312.method_1601(class_312.java:94)
        at net.minecraft.class_312.method_22686(class_312.java:165)
        at net.minecraft.class_1255.execute(class_1255.java:101)
        at net.minecraft.class_312.method_22684(class_312.java:165)
        at org.lwjgl.glfw.GLFWMouseButtonCallbackI.callback(GLFWMouseButtonCallbackI.java:36)
        at org.lwjgl.system.JNI.invokeV(Native Method)
        at org.lwjgl.glfw.GLFW.glfwWaitEventsTimeout(GLFW.java:3174)
        at com.mojang.blaze3d.systems.RenderSystem.limitDisplayFPS(RenderSystem.java:190)
        at net.minecraft.class_310.method_1523(class_310.java:1140)
        at net.minecraft.class_310.method_1514(class_310.java:728)
        at net.minecraft.client.main.Main.main(Main.java:217)
        at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:480)
        at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74)
        at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23)


A detailed walkthrough of the error, its code path and all known details is as follows:
---------------------------------------------------------------------------------------

-- Head --
Thread: Render thread
Stacktrace:
        at net.minecraft.class_310.method_29610(class_310.java:2087)
        at net.minecraft.class_310.method_29606(class_310.java:1939)
        at net.minecraft.class_528$class_4272.method_20174(class_528.java:437)
        at net.minecraft.class_528$class_4272.method_20164(class_528.java:343)
        at net.minecraft.class_528$class_4272.method_25402(class_528.java:279)
        at net.minecraft.class_350.method_25402(class_350.java:359)
        at net.minecraft.class_4069.method_25402(class_4069.java:27)
        at net.minecraft.class_312.method_1611(class_312.java:94)

-- Affected screen --
Details:
        Screen name: net.minecraft.class_526
Stacktrace:
        at net.minecraft.class_437.method_25412(class_437.java:490)
        at net.minecraft.class_312.method_1601(class_312.java:94)
        at net.minecraft.class_312.method_22686(class_312.java:165)
        at net.minecraft.class_1255.execute(class_1255.java:101)
        at net.minecraft.class_312.method_22684(class_312.java:165)
        at org.lwjgl.glfw.GLFWMouseButtonCallbackI.callback(GLFWMouseButtonCallbackI.java:36)
        at org.lwjgl.system.JNI.invokeV(Native Method)
        at org.lwjgl.glfw.GLFW.glfwWaitEventsTimeout(GLFW.java:3174)
        ```

Fourmisain commented:
> Test launching 1.17 (not with the try feature, just launching the game with a world):
> NoClassDefFoundError: Could not initialize class net.minecraft.class_2539

Uh, what?
I used the same version to test the build - did you use the latest branch?

`class_2539` = `NetworkState` or `NetworkPhase` (which we don't target at all...?)
The error apparently means a static initializer failed.

**Edit**: fixed

oliwoli commented:
tests with try feature:
1.17.1 :heavy_check_mark:
1.18.2 :heavy_check_mark: 
1.19.2 :heavy_check_mark: 
1.19.4-pre ❌ (shit... But it fails in the main branch too!)
1.19.4 ✔️
1.20.0 ✔️
1.20.1 ✔️
1.20.2 ✔️
1.20.3 ✔️
1.20.4 ✔️
1.20.5 ✔️
1.20.6 ✔️
1.21.0 ✔️
1.21.1 ✔️
1.21.2 ✔️
1.21.3 ✔️
1.21.4 ✔️
1.21.5-pre2 ✔️

Join Multiplayer Server Test:
1.20.3 ✔️
1.20.5 ✔️
1.21.1 ✔️
1.21.4 ✔️

Fourmisain commented:
Squashed most commits and wrote a slightly better commit message, should still be 1-to-1 be the same.

I guess this is ready.

oliwoli commented:
> Squashed most commits and wrote a slightly better commit message, should still be 1-to-1 be the same.
> 
> I guess this is ready.

@fiws permission to merge?

fiws commented:
LGTM. you can merge if it's ready.
